crates/astria-sequencer-relayer/src/relayer/submission.rs

```diff
@@ -0,0 +1,458 @@
+    }
+
+    fn from_path<P: AsRef<Path>>(path: P) -> eyre::Result<Self> {
+        let file = std::fs::File::open(&path)
+            .wrap_err("failed opening provided file path for reading post-submission state")?;
+        let state = serde_json::from_reader(file)
+            .wrap_err("failed reading contents of post-submission file")?;
+        Ok(state)
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "state")]
+enum PreSubmission {
+    Ignore,
+    Started {
+        sequencer_height: u64,
+        last_submission: PostSubmission,
+    },
+}
+impl PreSubmission {
+    fn from_path<P: AsRef<Path>>(path: P) -> eyre::Result<Self> {
+        let file = std::fs::File::open(&path)
+            .wrap_err("failed opening provided file path for reading pre-submission state")?;
+        let state = serde_json::from_reader(file)
+            .wrap_err("failed reading contents of pre-submission file")?;
+        Ok(state)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) struct SubmissionState {
+    pre: PreSubmission,
+    post: PostSubmission,
+    pre_path: PathBuf,
+    post_path: PathBuf,
+}
+
+#[derive(Debug)]
+pub(super) struct Started(SubmissionState);
+
+impl Started {
+    pub(super) fn finalize(self, celestia_height: u64) -> eyre::Result<SubmissionState> {
+        let Self(SubmissionState {
+            pre,
+            pre_path,
+            post_path,
+            ..
+        }) = self;
+
+        let PreSubmission::Started {
+            sequencer_height, ..
+        } = pre
+        else {
+            panic!(
+                "once intialized, a submission's `pre` field must always be `Started`. Here it is \
+                 not. This is a bug"
+            );
+        };
+
+        let new = SubmissionState {
+            pre,
+            post: PostSubmission::Submitted {
+                celestia_height,
+                sequencer_height,
+            },
+            pre_path,
+            post_path,
+        };
+
+        debug!(
+            state = serde_json::to_string(&new.post).expect("type contains no non-ascii keys"),
+            "finalizing submission by writing post-submit state to file",
+        );
+        let f = std::fs::File::options()
+            .write(true)
+            .truncate(true)
+            .open(&new.post_path)
+            .wrap_err("failed opening post-submit file for writing state")?;
+        serde_json::to_writer(&f, &new.post)
+            .wrap_err("failed writing post-submit state to file")?;
+        f.sync_all()
+            .wrap_err("failed syncing post-submit file to disk")?;
+        Ok(new)
+    }
+}
+
+impl SubmissionState {
+    pub(super) fn last_fetched_height(&self) -> Option<u64> {
+        match self.post {
+            PostSubmission::Fresh => None,
+            PostSubmission::Submitted {
+                sequencer_height, ..
+            } => Some(sequencer_height),
+        }
+    }
```

Comment:
last fetched height or last submitted height?

Reply:
Yes, thank you! I started getting confused with the naming here. I renamed this to `last_submitted_height` and also replaced the `u64` by `SequencerHeight` so that this already fails when parsing the file (not so important for the celestia height counterpart because `blob.Submit -> u64` and it's just for reporting).